### PR TITLE
[IOTDB-3296] ext-pipe suport .mods file.

### PR DESCRIPTION
**Pull request by jamber001**
https://issues.apache.org/jira/browse/IOTDB-3296

sync data ext-pipe fwk support use  .mods  files to filter deleted data.

### CODE REVIEW

**Comment by jamber001** [comment]
@HTHou  @qiaojialin  Could you help arrange reviewers for this PR?   :-)

**Comment by jamber001** [comment]
@qiaojialin @HTHou 
This PR is data-sync ext-pipe framework's consequent dev work.
Could you help arrange reviewers for this PR? :-)

File: server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    // == find existing overlap entries and remove them
+    Map<Long, Long> overlapEntries = delIntervalMap.subMap(startTime, true, endTime, true);
+    Iterator<Map.Entry<Long, Long>> iter = overlapEntries.entrySet().iterator();
+    while (iter.hasNext()) {
+      iter.next();
+      iter.remove();
+    }
+    //    Long[] keyList = overlapEntries.keySet().toArray(new Long[0]);
```
Comment (Cpaulyz) [inline server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java:129]: delete it

File: server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+  // TreeMap: StartTime => EndTime
+  private TreeMap<Long, Long> delIntervalMap;
+
+  public enum DeletedType {
+    NO_DELETED,
+    PARTIAL_DELETED,
+    FULL_DELETED
+  }
+
+  public static class TsInterval {
```
Comment (Cpaulyz) [inline server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java:56]: It seems that this internal class is not used？

File: server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    return delIntervalMap.isEmpty();
+  }
+
+  /**
+   * Check the deletion-state of the data-points of specific time range according to the info of
+   * .mods
+   *
+   * @param startTime - the start time of data set, inclusive
+   * @param endTime - the end time of data set, inclusive
+   * @return
```
Comment (Cpaulyz) [inline server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java:152]: Please add simple annotation for return value or `DeletedType`

File: server/src/test/java/org/apache/iotdb/db/sync/datasource/PipeOpManagerTest.java [hunk 4]
```diff
@@ -181,27 +241,160 @@ public void testPipSrcManager() throws IOException {
...
+    assertEquals("442", insertOperation.getDataList().get(i).right.get(0).getValue().toString());
+
+    i = 9;
+    assertEquals(
+        "root2.lemming.device3.sensor3", insertOperation.getDataList().get(i).left.getFullPath());
+    assertEquals(1617206403004L, insertOperation.getDataList().get(i).right.get(0).getTimestamp());
+    assertEquals("443", insertOperation.getDataList().get(i).right.get(0).getValue().toString());
+
+    // == test commit
+    //    pipeOpManager.commitData(sgName1, count1 - 5);
```
Comment (Cpaulyz) [inline server/src/test/java/org/apache/iotdb/db/sync/datasource/PipeOpManagerTest.java:396]: Uncomment these codes

**Review by Cpaulyz — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java (hunk 1 above)
Comment (jamber001) [reply server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java:129]: Thank your comment. I delete it.

File: server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java (hunk 3 above)
Comment (jamber001) [reply server/src/main/java/org/apache/iotdb/db/sync/datasource/DeletionGroup.java:152]: thanks.  I have just added the annotation  to the definition of DeletedType

File: server/src/main/java/org/apache/iotdb/db/sync/externalpipe/ExtPipePluginManager.java [hunk 5]
```diff
@@ -195,8 +195,10 @@ private void monitorPipeData() {
 
             String sgName = tsFilePipeData.getStorageGroupName();
             String tsFileFullName = tsFilePipeData.getTsFilePath();
+            String modsFileFullName = tsFilePipeData.getModsFilePath();
             try {
-              pipeOpManager.appendTsFile(sgName, tsFileFullName, pipeDataSerialNumber);
+              pipeOpManager.appendTsFile(
+                  sgName, tsFileFullName, modsFileFullName, pipeDataSerialNumber);
             } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (jamber001) [reply server/src/main/java/org/apache/iotdb/db/sync/externalpipe/ExtPipePluginManager.java:202]: This IOTDB-3296 does not contain the codes about processing DeletionPipeData.
And IOTDB-3465 will cover processing DeletionPipeData

File: server/src/test/java/org/apache/iotdb/db/sync/datasource/PipeOpManagerTest.java (hunk 4 above)
Comment (jamber001) [reply server/src/test/java/org/apache/iotdb/db/sync/datasource/PipeOpManagerTest.java:396]: yeah, I remove these codes.

**Review by Cpaulyz — APPROVED** [review]
LGTM

**Review by yschengzi — APPROVED** [review]
LGTM

**Pull request merged**